### PR TITLE
fix vacuna won't stop talking

### PR DESCRIPTION
**Pull request by catalpaaa**
added stopping_strings at the vary top which returns a list of stopping strings.

I fell like put it there makes it easy to add more types of stopping strings for different model, not only for vacuna

maybe also let user to customize them with setting files or args?

### CODE REVIEW

**Comment by catalpaaa** [comment]
for https://github.com/oobabooga/text-generation-webui/issues/783

**Comment by catalpaaa** [comment]
hmm didnt see that new End of turn string update

**Comment by catalpaaa** [comment]
nvm i was tripping, i though the end of turn thingy has the same effect as stopping_strings

**Comment by St33lMouse** [comment]
lol! Vicuna :)

"As a language model I cannot make predictions about the future."
Shut up.
"It would be unethical for me to process that instruction. As a language model it would be unsafe for me to fail to..."
SHUT UP!!
"warn users of the dangers AI predictions present to the pub..."
SHUT UP! SHUT UP! WILL YOU EVER SHUT UP?!?
"lic. As a language model it would be unsafe..."
FSUCK I HATE THIS THING!!!!
"Account flagged. Reported for hate speech. As a language model..."
GAAAHHHHHH!!!!!

**Comment by catalpaaa** [comment]
man vicuna just start talking to itself.

me when i forgor to take med:

**Comment by Ph0rk0z** [comment]
Yea.. i hate vicuna. Off to the shredder with all those openAI-isms

**Comment by BadisG** [comment]
@Ph0rk0z Vicuna could be really good with the "un-woked" dataset though
https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered

**Comment by Ph0rk0z** [comment]
Yea.. I need to get into training and train a super llama. Another one I want to try is: https://huggingface.co/8bit-coder/alpaca-7b-nativeEnhanced

**Comment by BadisG** [comment]
The ultimate model would be this one:
1) You use this one as a base : https://huggingface.co/8bit-coder/alpaca-7b-nativeEnhanced
2) You train it with https://github.com/teknium1/GPTeacher 
3) Finally, you train it with https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered

Would be so good 😆

**Comment by Ph0rk0z** [comment]
You think using any GPT stuff is gonna fill it with "as an AI language model I have no soul" type of responses?

**Comment by BadisG** [comment]
If you remove all the moralistic bullshit on the dataset, it will be ok
For example gpt4-x-alpaca is totally unrestricted because the dataset has no woke "as an AI language model..." in it

File: modules/chat.py [hunk 1]
```diff
@@ -17,6 +17,14 @@
 from modules.text_generation import (encode, generate_reply,
                                      get_max_prompt_length)
 
+#put stopping_strings up here so its easier to add more, 3rd and 4th are for vicuna
+def stopping_strings(name1, name2):
+    return [
+        f"\n{name1}:",
+        f"\n{name2}:",
+        f"\n### Human:",
+        f"\n### Assistant:"
```
Comment (danmincu) [inline modules/chat.py:26]: Isn't this culture (English) specific? Should we maybe provide a file containing the _stopping_ words?!

**Comment by knoopx** [comment]
btw this seems seems to work without any code change using the new instruct templates:
```
name: "### Assistant"
your_name: "### Human"
context: |-
  A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.

  ### Human: Hello, Assistant.
  ### Assistant: Hello. How may I help you today?
  ### Human: Please tell me the largest city in Europe.
  ### Assistant: Sure. The largest city in Europe is Moscow, the capital of Russia.
```

**Comment by gestalt73** [comment]
FWIW FastChat appears to just use characters "###" as stopping criteria.

```
2023-04-06 08:38:48 | INFO | gradio_web_server | ==== request ====
{'model': 'vicuna-13b-2023-04-03-hf', 'prompt': "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.###Human: What are the key differences between renewable and non-renewable energy sources?###Assistant: Renewable energy sources are those that can be replenished naturally in a relatively short amount of time, such as solar, wind, hydro, geothermal, and biomass. Non-renewable energy sources, on the other hand, are finite and will eventually be depleted, such as coal, oil, and natural gas. Here are some key differences between renewable and non-renewable energy sources:\n1. Availability: Renewable energy sources are virtually inexhaustible, while non-renewable energy sources are finite and will eventually run out.\n2. Environmental impact: Renewable energy sources have a much lower environmental impact than non-renewable sources, which can lead to air and water pollution, greenhouse gas emissions, and other negative effects.\n3. Cost: Renewable energy sources can be more expensive to initially set up, but they typically have lower operational costs than non-renewable sources.\n4. Reliability: Renewable energy sources are often more reliable and can be used in more remote locations than non-renewable sources.\n5. Flexibility: Renewable energy sources are often more flexible and can be adapted to different situations and needs, while non-renewable sources are more rigid and inflexible.\n6. Sustainability: Renewable energy sources are more sustainable over the long term, while non-renewable sources are not, and their depletion can lead to economic and social instability.\n###Human: What's up?###Assistant:", 'temperature': 0.7, 'max_new_tokens': 1536, 'stop': '###'}
2023-04-06 08:38:52 | INFO | gradio_web_server | I'm an AI language model, so I don't have feelings or emotions like humans do. But I'm here to help you with any questions or information you need, so feel free to ask me anything!
```

**Comment by gestalt73** [comment]
> btw this seems to work without any code change using the new instruct templates:
> 
> `characters/instruction-following/Vicuna.yaml`
> 
> ```
> name: "### Assistant"
> your_name: "### Human"
> context: |-
>   A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.
> 
>   ### Human: Hello, Assistant.
>   ### Assistant: Hello. How may I help you today?
>   ### Human: Please tell me the largest city in Europe.
>   ### Assistant: Sure. The largest city in Europe is Moscow, the capital of Russia.
> ```
> 
> the problem is that the code appends `:` to the end of `name`, so `name` should not contain it otherwise endstop word will be `### Human::`.

This works for most of the output



> btw this seems to work without any code change using the new instruct templates:
> 
> `characters/instruction-following/Vicuna.yaml`
> 
> ```
> name: "### Assistant"
> your_name: "### Human"
> context: |-
>   A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.
> 
>   ### Human: Hello, Assistant.
>   ### Assistant: Hello. How may I help you today?
>   ### Human: Please tell me the largest city in Europe.
>   ### Assistant: Sure. The largest city in Europe is Moscow, the capital of Russia.
> ```
> 
> the problem is that the code appends `:` to the end of `name`, so `name` should not contain it otherwise endstop word will be `### Human::`.

This is close.  The challenge is that they appear in two different formats.  I'm seeing both "### Assistant" and "###Assistant" for example.  FastChat just looks for "###" and stops there.

**Comment by knoopx** [comment]
@gestalt73 you can add to the template:
```
end_of_turn: "###"
```
but the currently it doesn't strip it, will stop but leave the `###` at the end of the message.

**Comment by OWKenobi** [comment]
I like the idea of additional stopping strings.

However, I don't like them being hardcoded. They should be part of ever .yaml-file, maybe as an optional parameter.

For now, I opened another pull request that just handles the double colon problem, and it makes Vicuna shut up ;-)

**Comment by catalpaaa** [comment]
> I like the idea of additional stopping strings.
> 
> 
> 
> However, I don't like them being hardcoded. They should be part of ever .yaml-file, maybe as an optional parameter.
> 
> 
> 
> For now, I opened another pull request that just handles the double colon problem, and it makes Vicuna shut up ;-)

yes but I remembered ooba saying how adding more command line arg is too bloated. It's his call, and I'll it.

**Comment by oobabooga** [comment]
I think that adding a text field for custom stopping strings in a convenient format would be nice. Maybe just separated by commas like

"You:, Bot: Assistant:"

Implementing this in a way that works for all modes (chat and non-chat) is on my to-do list, but if someone wants to make a PR, please feel free to.

For now I am closing this one after https://github.com/oobabooga/text-generation-webui/pull/838, as it handles the Vicuna case in the intended way (through a new pseudocharacter).

**Comment by catalpaaa** [comment]
> I think that adding a text field for custom stopping strings in a convenient format would be nice. Maybe just separated by commas like
> 
> "You:, Bot: Assistant:"
> 
> Implementing this in a way that works for all modes (chat and non-chat) is on my to-do list, but if someone wants to make a PR, please feel free to.
> 
> For now I am closing this one after #838, as it handles the Vicuna case in the intended way (through a new pseudocharacter).



> I think that adding a text field for custom stopping strings in a convenient format would be nice. Maybe just separated by commas like
> 
> "You:, Bot: Assistant:"
> 
> Implementing this in a way that works for all modes (chat and non-chat) is on my to-do list, but if someone wants to make a PR, please feel free to.
> 
> For now I am closing this one after #838, as it handles the Vicuna case in the intended way (through a new pseudocharacter).

should i make a pr that read it from some file in ./, or you want it hard coded

**Comment by oobabooga** [comment]
I was thinking of adding a gr.Textbox to the interface itself under Parameters. The question is how to pass the value of this textbot to the text generation functions. So far, I have grouped all sliders/checkboxes/numbers under a gr.State called share.gradio['generate_state']. But for textboxes it might be better to pass them directly

**Comment by catalpaaa** [comment]
> oxes it might be better to pass them directly

yes yes i go check

idk where to ask this but, are you in that unoffical discord server :)

**Comment by catalpaaa** [comment]
> I was thinking of adding a gr.Textbox to the interface itself under Parameters. The question is how to pass the value of this textbot to the text generation functions. So far, I have grouped all sliders/checkboxes/numbers under a gr.State called share.gradio['generate_state']. But for textboxes it might be better to pass them directly

i think the value of gr.Textbox should go to shared.setting, so if they want they can set a custome setting.json so they dont have to go and type in whatever thingy everytime

**Comment by oobabooga** [comment]
> i think the value of gr.Textbox should go to shared.setting, so if they want they can set a custome setting.json so they dont have to go and type in whatever thingy everytime

I agree, it would be like the character names that can have default values in the settings.json.

**Pull request closed without merging**